sycl/source/detail/graph_impl.hpp

```diff
@@ -585,13 +610,35 @@ class partition {
       MPiCommandBuffers;
   /// List of predecessors to this partition.
   std::vector<std::shared_ptr<partition>> MPredecessors;
+  /// True if the graph of this partition is a single path graph
+  /// and in-order optmization can be applied on it.
+  bool MIsInOrderGraph = false;
 
   /// @return True if the partition contains a host task
   bool isHostTask() const {
     return (MRoots.size() && ((*MRoots.begin()).lock()->MCGType ==
                               sycl::detail::CG::CGTYPE::CodeplayHostTask));
   }
 
+  /// Checks if the graph is single path, i.e. each node has a single successor.
+  /// If so, the MIsInOrderGraph flag is set.
+  void checkIfGraphIsSinglePath() {
+    MIsInOrderGraph = true;
```

Comment:
I was thinking about leaving this comment in internal review, but didn't quite - I think this function could be less stateful by returning a `bool` rather than setting `MIsInOrderGraph` internally, and then the call site should be `MIsInOrderGraph = checkIfGraphIsSinglePath()`

Reply:
I've made this change 👍